### ISSUE TITLE
Optimize OpenBLAS for LinuxOne s390x zEC13 SIMD: BLAS 1/2 single precision

### ISSUE THREAD

**Issue by edelsohn** [xianyi/OpenBLAS#992]

**Comment by papaaloabob** [comment]
Aloha edelsohn,
Is this still open?  I see that quickwritereader did a lot of related work.  Is he not interested in this one?

Thank you,
bob

**Comment by quickwritereader** [comment]
As there is not direct hardware simd I postponed it. as for some functions I believed using simd will not improve performance much.  So I decided to take other issue before returning to this one. 
I'm thinking to try it in near future. 
But If you want to try go ahead. After all solving this issue will benefit community.

**Comment by edelsohn** [comment]
There is Power VSX SIMD support for single precision, so there should be opportunities for optimization and tuning. It may be best to implement in C with vector types than directly using intrinsics. One needs to experiment.

**Comment by papaaloabob** [comment]
Thank you very much for your quick and thoughtful reply.  I'll look closer.

Bob Holmes
808-640-1799

**Comment by mtk-i** [comment]
The processors in [IBM LinuxONE](https://developer.ibm.com/linuxone) can't handle SIMD single precision instructions, I have successfully assembled SIMD single precision instructions but I got it `Illegal instruction` when I run it. It's good to point me to a capable Z14 VM since I ported many functions to single precision so I can test it. Thanks in advance.

**Comment by edelsohn** [comment]
Unfortunately, there are no publicly accessible z14 systems. The best that I can do is test on a z14 system internal to IBM and provide feedback.

**Comment by mtk-i** [comment]
Some functions need many tests to get optimal performance, However I will do my best to make it optimized as much as possible and send it to you, Maybe I will provide many implementations for some functions, Do you have problem in that?

**Comment by mtk-i** [comment]
hmm I have an idea, since I can compile the instructions, I will write separated unit tests that compatible with  functions implementations and just send the execution files to you for test.

**Comment by edelsohn** [comment]
I don't know if QEMU for s390x is up to date yet for these instructions. That's another option.

**Comment by quickwritereader** [comment]
If these additions are for z13 I could share the pc given to me.

**Comment by mtk-i** [comment]
@edelsohn Not sure if QEMU give native performance, Although I do not think so. I think providing source code to you "not execution files since can't be trusted by IBM" for test should work. Let's see how we can test it later.
@quickwritereader unfortunately it's z14 specific, Z13 can't handle that.

**Comment by mtk-i** [comment]
Done
To get optimal performance of single precision functions I emulated the implementation approach using double precision instructions so I concluded in one implementation for each function, So after intense tests I concluded that @quickwritereader come with the better approaches for double precision functions, Some single precision functions followed the same approach, Others have been even more tuned before the porting. There are exceptions for a few functions. However the whole ZARCH functions have been reconstructed and the reason for most cases to keep single precision and double precision functions consistent. 

Features:
-BLAS 1/2 single precision implementations except nrm2.
-missing BLAS 1/2 double precision implementations except nrm2.
-Gemv optimization, After optimize Zgemv I noticed bug in zgemv_n_4 and zgemv_t_4 C implementation so I decided to upload the other functions to review and test them while I take a look to the problem.
-Add special unit tests so test and benchmark functions can done easily.

How to test:
-Get Z14 branch [here ](https://github.com/maamountki/OpenBLAS/tree/z14)
-build OpenBLAS
-build ztest
-Test. e.g. `./sgemv.goto 1000 1002` and report back if something failed.

**Comment by martin-frbg** [comment]
Not sure why you removed the "detect" function rather than adding the machine numbers for z14 (which appear to be 3906, 3907) ? Your branch also removes the zgemv_n_4 and zgemv_t_4 implementations from z13, so you seem to think there is an error already in #1499 ?

**Comment by quickwritereader** [comment]
@maamountki well done. that's indeed great contribution
I noticed that you added missing min,max and also ported some  vector codes  into  pure inline assembly ,too. 
As for nrm2 I believe there is not need to implement it. Because precision matters more than speed.
I also noticed that you preferred mvc option of copy. I believe it was stable one. but somehow I chosed raw vl vst pairs and put mvc inside ifdef section, I forgot exact reason.
(
I'd like also share my recent views on kernel:
since power8 I started to use named arguments which makes changing inline assembly a lot easier.  I think it would be better to refactor z13/14 inline assemblies too.
Also tripling or doubling loop internals to make reordering easier. That one I tried on z13 too but it did not yield any improvement. Though still good technique
)

**Comment by quickwritereader** [comment]
strange gemv kernels were written c simd vectors and I did not use any inline assembly.  I think if there you found any bug we could solve it easily

**Comment by mtk-i** [comment]
@martin-frbg I was working on old branch where "detect" not added yet, I will add it now, And for zgemv_n_4 and zgemv_t_4 I get wrong but close results in the pure C implementation, It's not related in SIMD.
@quickwritereader About nrm2 I also expected that "every element is related to previous one as well as a branch inside the loop". And actually I tested all your optional implementations and removed the slower ones, and yes mvc was faster. Also I will take a look to the arguments.

**Comment by quickwritereader** [comment]
I just checked zgemv with BLAS-Tester. It did not complain. (Own fork)

**Comment by mtk-i** [comment]
I checked it again, It's failing even with vector kernels undefined.

**Comment by quickwritereader** [comment]
I don't have any idea. Could you check my fork,too. 
thanks

**Comment by mtk-i** [comment]
k, If you sure that zgemv_n_4 and zgemv_t_4 work correctly so perhaps the bug in C implementation in arm directory, Let me do more investigating.

**Comment by mtk-i** [comment]
Nevermind, there are no bug at all, I just forgot the scale stuff. However I create [#1723](https://github.com/xianyi/OpenBLAS/pull/1723) PR to disable scale by default.

**Comment by martin-frbg** [comment]
Not sure if we need to change a pre-existing benchmark just because someone got bitten trying to use it as a test for code correctness without checking the arguments ? (Not a big deal either way though). @maamountki xianyi's BLAS tester code at https://github.com/xianyi/BLAS-Tester is probably a better option for this.

**Comment by mtk-i** [comment]
The idea behind ztest was to support multiple BLAS implementations so we can pick the better one after test, When I ended up in one implementation for each function I decided to keep it in my branch since I'm comfortable with it. Anyway I believe the author of "pre-existing benchmark" didn't mean to benchmark zgemv with scale while benchmark gemv without scale regardless it's holy code or not that's not consistent.

**Comment by martin-frbg** [comment]
Had not looked at the actual benchmark to realize that it is one source for all xGEMV. Agree now that it looks like a copy-paste of the declaration of alpha that went unnoticed.

**Comment by mtk-i** [comment]
I added Zgemv to the branch. @edelsohn you can check it out.

**Comment by edelsohn** [comment]
Excellent!  Great progress!

**Comment by martin-frbg** [comment]
Things went a bit quiet here it seems - is this ready for merging, or still waiting for someone from IBM to verify on actual hardware ?

**Comment by quickwritereader** [comment]
as far as I know those works were done for z14. so the issue should be still here as it was created for z13.
I recently was working on power8 blas1 single . then I had to stop working because of intensity in my courses. I'm thinking of to restart power8 works.

**Comment by mtk-i** [comment]
It isn't ready for merging, I need to test it first but [LinuxONE™ Virtual Machines](https://developer.ibm.com/linuxone/account-vm-login) still doesn't support Z14. @edelsohn does [IBM Cloud Private](https://developer.ibm.com/linuxone/account-icp-login) support Z14?

**Comment by edelsohn** [comment]
I was confused about the status as well. I have deferred to @quickwritereader recently and there seemed to be some disagreement about unreproducible failures last August.

I didn't realize that the difference in results had been resolved and the question was testing on a z14 system. LinuxONE Community Cloud still does not support z14. I will test this on a z14 system next week.

**Comment by edelsohn** [comment]
My colleague tested the patch (z14 branch) and discovered problems. He fixed the build problems but the code segfaults when run. I will let him provide more details.

**Comment by aarnez** [comment]
> My colleague tested the patch (z14 branch) and discovered problems. He fixed the build problems but the code segfaults when run. I will let him provide more details

Right. Using the "z14" branch from [here](https://github.com/maamountki/OpenBLAS.git), I added this patch to fix the build:
[0001-Fix-z14-build.txt](https://github.com/xianyi/OpenBLAS/files/2725083/0001-Fix-z14-build.txt)
With that the build stops when hitting a segfault with this command:
~~~
OPENBLAS_NUM_THREADS=1 OMP_NUM_THREADS=1 ./sblat2 < ./sblat2.dat
~~~
The command yields:
~~~
Program received signal SIGSEGV: Segmentation fault - invalid memory reference.

Backtrace for this error:
#0  0x3fffdd9be11 in ???
#1  0x3fffdd9ae39 in ???
#2  0x3ffffff0aa5 in ???
#3  0x103be76 in sgemv_kernel_4x4
	at ../kernel/zarch/sgemv_t_4.c:34
#4  0x103cc7f in sgemv_t
	at ../kernel/zarch/sgemv_t_4.c:577
#5  0x103e725 in ssymv_U
	at ../kernel/zarch/../generic/symv_k.c:78
#6  0x10182bf in ssymv_
	at /home/arnez/OpenBLAS/interface/symv.c:188
#7  0x100c949 in schk2_
	at /home/arnez/OpenBLAS/test/sblat2.f:907
#8  0x1016457 in sblat2
	at /home/arnez/OpenBLAS/test/sblat2.f:311
#9  0x1016be9 in main
	at /home/arnez/OpenBLAS/test/sblat2.f:390
Segmentation fault (core dumped)
~~~
@maamountki, do you need any more information for debugging this issue?

**Comment by mtk-i** [comment]
I will take a look and let u know if I will need more info.

**Comment by mtk-i** [comment]
I committed a fix for sgemv_t_4, could u test with that fix.

**Comment by aarnez** [comment]
> I committed a fix for sgemv_t_4, could u test with that fix.

Sure, thanks for the fix. With that I now get
~~~
 SGEMV  PASSED THE TESTS OF ERROR-EXITS

 ******* FATAL ERROR - COMPUTED RESULT IS LESS THAN HALF ACCURATE *******
           EXPECTED RESULT   COMPUTED RESULT
       1      0.240646         -0.542240E-01
       2     -0.217188         -0.238480    
 ******* SGEMV  FAILED ON CALL NUMBER:
    584: SGEMV ('T',  4,  2, 1.0, A,  5, X, 1, 0.0, Y, 1)         .
~~~
and
~~~
 SSYMV  PASSED THE TESTS OF ERROR-EXITS

 ******* FATAL ERROR - COMPUTED RESULT IS LESS THAN HALF ACCURATE *******
           EXPECTED RESULT   COMPUTED RESULT
       1       1.14768           1.14768    
       2      0.618986          0.618986    
       3       1.09204           1.09204    
       4     -0.227969         -0.227969    
       5      -1.22627          -1.22627    
       6      0.897953          0.897953    
       7     -0.886102         -0.886102    
       8     -0.931212         -0.931212    
       9      0.505439          0.505439    
      10      0.668288          0.668288    
      11      0.957609          0.957609    
      12      -1.42113          -1.42113    
      13     -0.417240         -0.417240    
      14       1.50052           1.50052    
      15     -0.308408         -0.308408    
      16     -0.841416         -0.841416    
      17      0.855837          0.859233    
      18     -0.969934         -0.437599E-01
      19      0.275873          0.497033    
      20     -0.413747          0.608308E-01
      21       1.52083           1.47286    
      22      0.236277          0.823904    
      23      0.135978E-01     -0.218410    
      24     -0.400299         -0.462971    
      25      0.159057         -0.827319E-01
      26      0.442964          0.645785    
      27      -1.82839          -2.19146    
      28      0.908607          0.676599    
      29      0.344885          0.906065    
      30      0.147954E-01      0.235781    
      31     -0.493687         -0.856933    
 ******* SSYMV  FAILED ON CALL NUMBER:
   1157: SSYMV ('U', 31, 1.0, A, 32, X, 1, 0.0, Y, 1)             .
~~~
and then
~~~
OPENBLAS_NUM_THREADS=1 OMP_NUM_THREADS=1 ./cblat2 < ./cblat2.dat

Program received signal SIGSEGV: Segmentation fault - invalid memory reference.

Backtrace for this error:
#0  0x3fffdd9be11 in ???
#1  0x3fffdd9ae39 in ???
#2  0x3ffffffbb3d in ???
#3  0x106abde in cgemv_kernel_4x2
	at ../kernel/zarch/cgemv_n_4.c:155
#4  0x106b3e5 in cgemv_n
	at ../kernel/zarch/cgemv_n_4.c:417
#5  0x107ae5b in chemv_U
	at ../kernel/zarch/../generic/zhemv_k.c:116
#6  0x101b513 in chemv_
	at /home/arnez/OpenBLAS/interface/zhemv.c:203
#7  0x100da93 in cchk2_
	at /home/arnez/OpenBLAS/test/cblat2.f:921
#8  0x10179e5 in cblat2
	at /home/arnez/OpenBLAS/test/cblat2.f:316
#9  0x1018291 in main
	at /home/arnez/OpenBLAS/test/cblat2.f:397
OPENBLAS_NUM_THREADS=1 OMP_NUM_THREADS=1 ./zblat3 < ./zblat3.dat
/bin/sh: line 1: 62479 Segmentation fault      (core dumped) OPENBLAS_NUM_THREADS=1 OMP_NUM_THREADS=1 ./cblat2 < ./cblat2.dat
~~~
~~~

**Comment by mtk-i** [comment]
I fixed the last problem and I'm trying to figure out what's wrong with SGEMV with no chance, could u please comment out lines 475 ... 477 in sgemv_t_4.c and try, Sorry for inconvenience.

**Comment by aarnez** [comment]
No problem. With the the conditional invocation of add_y_kernel_4 commented out, the result looks the same as before.

**Comment by mtk-i** [comment]
SGEMV implementation looks ok that's weird, However I created 2 different implementations one modifies a few instructions and the other have different approach. [1-sgemv_t_4.txt](https://github.com/xianyi/OpenBLAS/files/2729206/sgemv_t_4.txt) [2-sgemv_t_4.txt](https://github.com/xianyi/OpenBLAS/files/2729207/sgemv_t_4.txt)

**Comment by quickwritereader** [comment]
I would like to share my experience how to test for 90% problems and correct problems step by step

1. I think its better to test with  C implementation  of future Assembly microkernels to see if it passes tests .  And check if there are warnings for uninitialized variables and so on in that kernel.
2. For testing microkernels with maximally avoiding inlining errors and strange results ensure that all functions have  `__attribute__ ((noinline))` added. At this step segfaults should be avoided
3. If microkernels have errors its better to use test programs inside ctest folder. For sgemv one can get copy  of sin2 and modify it to test for only sgemv with exact m n's ( for example 4x4 8x4 4x2 16x16) to identify if parts of kernel works properly and check if the problem arise while using two or three of them together. Note that its better to test with single thread
```
export OMP_NUM_THREADS=1 #test only with single thread   
./xscblat2 <sin2_copy  #for not changing original sin2 test using copy file
```
There is also [BLAS-TESTER](https://github.com/xianyi/BLAS-Tester) program which can make life easier to find such problems. 

> to reduce compilation time (without calling  make clean) one can remove from kernel folder exact .o object files of those kernels and compile again.

4. If one wants to let compiler decide inlining then `__attribute__ ((noinline))  `could be removed. If after removing some errors occur then one must solve clobbering memory issues

**Comment by mtk-i** [comment]
@quickwritereader thanks it makes sense to do these steps to catch up the problem, However I tested the C implementation and tests passed, also I can determine that sgemv_kernel_4x2 caused the problem from @aarnez summary report, and yes it's worth a try with `__attribute__ ((noinline))`

**Comment by aarnez** [comment]
> SGEMV implementation looks ok that's weird,

After writing to an FPR with an instruction like "aebr" you relied on other elements of the corresponding vector register to remain intact. This is not supported by the architecture: _"Whenever a floating-point instruction or floating point support instruction writes to a floating point register, or a floating point instruction that reads a register pair reads from floating-point registers, bits 64-127 of the corresponding vector register are unpredictable."_

> However I created 2 different implementations one modifies a few instructions and the other have different approach. [1-sgemv_t_4.txt](https://github.com/xianyi/OpenBLAS/files/2729206/sgemv_t_4.txt) [2-sgemv_t_4.txt](https://github.com/xianyi/OpenBLAS/files/2729207/sgemv_t_4.txt)

That second version, 2-sgemv_t_4.txt, fixes various occurrences of the problem described above. With that plus a small typo fix the build progresses further and then runs into this:
~~~
 ******* FATAL ERROR - COMPUTED RESULT IS LESS THAN HALF ACCURATE *******
                       EXPECTED RESULT                    COMPUTED RESULT
       1  (  -0.607192    ,  -0.262085    )  (  -0.420868E-01,  -0.295458    )
       2  (   0.180449    ,  -0.434689    )  (   0.335272    ,  -0.355538    )
       3  (   0.384241    ,  -0.191735    )  (   0.460220    ,   0.212586    )
       4  (  -0.485153    ,   0.185071    )  (   0.149114    ,   0.400899    )
       5  (  -0.223453E-01,  -0.546696E-01)  (   0.459942    ,   0.584042E-01)
       6  (  -0.335252    ,   0.408274    )  (   0.128284    ,   0.183005    )
       7  (   0.416776    ,  -0.864290    )  (   0.771882    ,  -0.247994E-01)
       8  (  -0.406477    ,  -0.696984    )  (  -0.429212    ,  -0.861106    )
       9  (   0.135872    ,   0.509391E-01)  (   0.135872    ,   0.509391E-01)
      10  (  -0.353782    ,   0.852136    )  (  -0.353782    ,   0.852136    )
      11  (  -0.962145E-01,   0.217275E-01)  (  -0.962145E-01,   0.217275E-01)
 ******* CGEMV  FAILED ON CALL NUMBER:
   2168: CGEMV ('N', 11,  7,( 1.0, 0.0), A, 12, X, 1,( 0.0, 0.0), Y, 1)         .
~~~
@maamountki, for the result above I used this patchset on top of your branch:
[fixes-for-z14.txt](https://github.com/xianyi/OpenBLAS/files/2737718/fixes-for-z14.txt).

**Comment by mtk-i** [comment]
I committed a fix for this test failure.

**Comment by aarnez** [comment]
Thanks, now all of the above succeeds. Next is a precision error with dsdot:
~~~
./openblas_utest
...
TEST 12/23 dsdot:dsdot_n_1 [FAIL]
  ERR: test_dsdot.c:47  expected -2.393e-03, got -2.393e-03 (diff -1.035e-10, tol 1.000e-13)
~~~
@maamountki , this may be due to the fact that your implementation uses single-precision multiply instructions (vfmsb). The reference implementation performs double-precision arithmetic only.

**Comment by mtk-i** [comment]
Could u test the new dsdot implementation in the branch.

**Comment by aarnez** [comment]
Good, another problem fixed!  Note that I also adjusted the C code in dsdot.c to cast everything to double first.
Now I see:
~~~
OPENBLAS_NUM_THREADS=2 ./xccblat2 < cin2
 TESTS OF THE COMPLEX          LEVEL 2 BLAS
...
 cblas_cgemv  PASSED THE COMPUTATIONAL TESTS (  4324 CALLS)
xccblat2: zgemv.c:274: cblas_cgemv: Assertion `stack_check == 0x7fc01234' failed.

Program received signal SIGABRT: Process abort signal.
~~~
This seems tricky. Let me know if you need more information.

**Comment by mtk-i** [comment]
Just to know where the problem, Could u set CGEMVNKERNEL = ../arm/zgemv_n.c in KERNEL.Z14

**Comment by aarnez** [comment]
With that setting the problem disappears, and the build completes!

**Comment by mtk-i** [comment]
Could u replace `buffer_size = (buffer_size + 3) & ~3;` with `buffer_size = (buffer_size + 15) & ~15;` in interface/zgemv.c

**Comment by aarnez** [comment]
That change makes no difference; the same assertion occurs.

**Comment by mtk-i** [comment]
Could u test the latest update of the branch.

**Comment by aarnez** [comment]
Thanks, this helps. The build completes now!

**Comment by edelsohn** [comment]
Great! If everything now builds, what is the next step?
Build means that OpenBLAS builds and runs or that the tests build, link against OpenBLAS and run?
How is the performance on z14? The issue does not require ideal performance, but I hope that it provides an initial boost.
What is the OpenBLAS community criteria for landing the patch?

**Comment by mtk-i** [comment]
Well, The kernels designed while ideal performance in mind. In theory single precision should be faster than double precision, You can test the performance and let me know if you suspect in a function. Once that done I will create pull request to this repository, BTW @aarnez u did some fixes into the overall patch, Do u want to create pull request that can be applied to my branch or I just push them to the branch?

**Comment by mtk-i** [comment]
Also during build process, OpenBLAS run a bunch of tests which I think it's enough to ensure correctness of the implementations.

**Comment by martin-frbg** [comment]
The tests are usually good for catching the most glaring bugs, but (as noted earlier) BLAS-Tester is a bit more comprehensive. Not sure if anybody will want to run Julia on Z14, which usually exposes all the hidden bugs ... In this case the changes seem to be confined to zarch so it is "probably" safer to merge than some other things in recent history.

**Comment by edelsohn** [comment]
There is some effort to port Julia to z14, but, again, that's not the goal of this issue.

**Comment by aarnez** [comment]
1.  BLAS-Tester results: All tests pass, except for the tests of two of the AMAX variants:

	~~~~
	$ ./xcl1blastst -R amax 2>&1 | grep ERROR
	ERROR:  resid=8388608.000000, tr_max=54, te_max=44, eps=1.192093e-06
	ERROR:  resid=91435824.000000, tr_max=114, te_max=5, eps=1.192093e-06
	ERROR:  resid=15099494.000000, tr_max=430, te_max=412, eps=1.192093e-06
	ERROR:  resid=435368768.000000, tr_max=675, te_max=156, eps=1.192093e-06
	ERROR:  resid=286890400.000000, tr_max=398, te_max=56, eps=1.192093e-06
	ERROR:  resid=263402288.000000, tr_max=323, te_max=9, eps=1.192093e-06
	$ ./xsl1blastst -R amax 2>&1 | grep ERROR
	ERROR:  resid=487378112.000000, tr_max=758, te_max=177, eps=1.192093e-06
	$ 
	~~~~

2.  Performance:  At first glance the asymptotic sgemv performance looks significantly better than the upstream version.  I haven't really compared anything else yet.  The patch obviously doesn't contain any speedups for sgemm.

3.  Next steps: First the AMAX problem above should be fixed.  Then, if it helps, I can create a pull request for the patches I've been using on top of maamountki's.  We probably want the z14 patches to be consolidated before pulling them into "develop", right?  Are there any formal requirements, such as formatting rules?

**Comment by mtk-i** [comment]
I will take a look at AMAX. SGEMM is BLAS level 3 function which not required for this issue, Could you test the performance of other Blas level 1/2 single precision functions btw you can see the performance of DGEMV and ZGEMV and some double precision max/min functions since I optimized them too.

**Comment by martin-frbg** [comment]
No special formatting rules or other formal requirements as far as I know (though perhaps check for mis-spelled comments or trailing whitespace so that we do not get any "boring" whitespace-only PRs later that touch dozens of files without actually improving the compiled library). Consolidating the patches would be nice just for clarity as well.
"not required for this issue" is perhaps a gray area as the original issue was about z13 and kind of got turned into "add support for z14" (though it seems some of your work affects z13 too if I read it correctly ?). Still I think BLAS3 improvements will be just as welcome later. :-)

**Comment by edelsohn** [comment]
We originally split the issues as BLAS 1/2 and BLAS 3.  This originally was z13 but was transformed to z14.  The original z13 BLAS 3 issue made sense with the z13 instruction to load/store a pair of single precision values. But hand-written BLAS 1/2 only made sense with z14 single precision instructions.

We can consider a z14 BLAS 3 issue once this is complete, if this was a productive approach.

**Comment by mtk-i** [comment]
z14 BLAS 3 support absolutely need z14 virtual machine, I see many cpu architectures here share the same BLAS 1/2 kernels across all cpu models and that make sense actually but when it comes to BLAS 3 the work should be done on the same model.

**Comment by edelsohn** [comment]
Yes, I completely agree.  Andreas and I are very pleasantly surprised at the progress you have made without access to hardware.  Adding z14 hardware to the LinuxONE Community Cloud is somewhere in the plan. Hardware access clearly is necessary for non-IBM developers to contribute to Open Source on IBM systems.

**Comment by quickwritereader** [comment]
If isamin or isamax not implemented I could provide power8 altivec version  afaik they could be easily converted to z14. From performance point it should at least triple or quadruple base c implementation.

**Comment by mtk-i** [comment]
@quickwritereader Thanks, isamin and isamax are already implemented, You can see them in the branch.

**Comment by quickwritereader** [comment]
that's great ( I saw AMAX and forgot to check your fork. )

**Comment by mtk-i** [comment]
@aarnez could you see if this help [samax.txt](https://github.com/xianyi/OpenBLAS/files/2760069/samax.txt)

**Comment by aarnez** [comment]
@maamountki: sorry, this patch doesn't change the output of "xcl1blastst -R amax" or "./xsl1blastst -R amax".

**Comment by mtk-i** [comment]
I committed a fix in the branch.

**Comment by aarnez** [comment]
Sorry, this doesn't change the result either.  It seems that BLAS-Tester only tests with an increment of 1 per default, so that's what the problem has been seen with. And samax and friends don't seem to be invoked by BLAS-Tester at all, but only isamax and friends.
While looking at samax.c I noticed that the inline assembly doesn't use VFMAX, which I thought would be a perfect fit here, or isn't it?  And in the C code I wonder whether the manual loop-unrolling really pays off.  In this case it probably increases the number of branches within the loop, which may not be ideal.

**Comment by mtk-i** [comment]
I fixed possible bugs in the max/min functions plus replace VFCH with VFMAX/VFMIN when possible.
Manual loop-unrolling in C for non single increment implementation improve the performance for these functions.

**Comment by mtk-i** [comment]
Does something go wrong with the last update? I'm just worried about it because I chose IEEE standard for adapting VFMAX/VFMIN so maybe it will cause conflict with C implementation in BLAS-Tester.

**Comment by aarnez** [comment]
Sorry for the delay...
With your latest update I've still been seeing the same problem in BLAS-Tester. I will look into the failures in more detail. In addition, I noticed that most of the min/max functions are not covered by the tests at all. Thus I wrote some tests myself. I'll probably have results early next week.

Whether the "IEEE" methods are adequate, I don't know. Without a formal definition of the BLAS min/max functions, the only "documentation" I've seen is the reference implementation. That one uses a comparison, making it asymmetric. It seems unlikely that this is intentional, thus a formal definition would probably not require the same behavior. IMHO the IEEE methods are more reasonable. In any case, the choice of method affects the result only if the array contains something other than regular finite numbers. Thus I don't think that the BLAS-Tester fails are related to this.
(If anyone has more information about the intended behavior of the BLAS min/max functions when the array contains NaNs, infinities, or negative zeroes, that would be great.)

**Comment by mtk-i** [comment]
I fixed bug in iamax/iamin/imax/imin hopefully this solve the test failure.

**Comment by mtk-i** [comment]
I just realized it's useless, sorry.

**Comment by mtk-i** [comment]
Now real fix has been pushed to my fork.

**Comment by edelsohn** [comment]
Apparently there will be a new Community Cloud system in South Africa running on a z14 system that should provide easier access and testing.  It's not available yet and I don't know how soon public accounts will be available.

**Comment by aarnez** [comment]
> Now real fix has been pushed to my fork.

Good! Now isamax doesn't yield a test failure any more, and there's just a problem with icamax/icamin left:
~~~
$ ./xcl1blastst -R amax 2>&1 | grep ERROR
ERROR:  resid=39426456.000000, tr_max=54, te_max=7, eps=1.192093e-06
ERROR:  resid=327994560.000000, tr_max=430, te_max=39, eps=1.192093e-06
ERROR:  resid=135056592.000000, tr_max=360, te_max=199, eps=1.192093e-06
ERROR:  resid=506671936.000000, tr_max=675, te_max=71, eps=1.192093e-06
ERROR:  resid=32715572.000000, tr_max=398, te_max=359, eps=1.192093e-06
ERROR:  resid=50331648.000000, tr_max=323, te_max=263, eps=1.192093e-06
ERROR:  resid=148478368.000000, tr_max=472, te_max=295, eps=1.192093e-06
$ ./xcl1blastst -R amin 2>&1 | grep ERROR
ERROR:  resid=69625448.000000, tr_min=21, te_min=104, eps=1.192093e-06
ERROR:  resid=156028112.000000, tr_min=590, te_min=776, eps=1.192093e-06
ERROR:  resid=254174816.000000, tr_min=345, te_min=648, eps=1.192093e-06
ERROR:  resid=75497472.000000, tr_min=46, te_min=136, eps=1.192093e-06
~~~

**Comment by mtk-i** [comment]
@aarnez this should be fixed now.

@edelsohn while this is great news I'm not sure if something need to be add when I put my hands on z14 instance. After fixing this problem the patch will be complete for both stability and performance IMO. Anyway It's all up to you and personally I didn't expect we will get this bunch of errors when testing so I feel sorry for taking much time from @aarnez

**Comment by aarnez** [comment]
> ... this should be fixed now.

Almost:
~~~
$ ./xcl1blastst -R amax 2>&1 | grep ERROR
ERROR:  resid=77175192.000000, tr_max=675, te_max=583, eps=1.192093e-06
~~~

> I didn't expect we will get this bunch of errors when testing so I feel sorry for taking much time from @aarnez

Don't worry. This is not a small project. Considering the amount of code, I'd actually have expected much more bugs than we've found so far.

**Comment by quickwritereader** [comment]
@maamountki @aarnez  Indeed he did great work. all these works without accessing real pc

**Comment by edelsohn** [comment]
Excellent work. Looks like support is really close to finished.

**Comment by mtk-i** [comment]
@aarnez should be fixed completely now.

**Comment by aarnez** [comment]
@maamountki, great! The BLAS-Tester failures are all gone now!

**Comment by edelsohn** [comment]
Success?

**Comment by joncmu** [comment]
@maamountki  One comment on performance for the new camax and camin functions.  vlef instructions are dependent on the last writer of the target register.  If you can afford 2 scratch registers you can do the loading much more efficiently.

So this sequence: 
        "vlef  %%v16,0(%%r1,%2),0         \n\t"
        "vlef  %%v17,4(%%r1,%2),0         \n\t"
        "vlef  %%v16,8(%%r1,%2),1         \n\t"
        "vlef  %%v17,12(%%r1,%2),1        \n\t"
        "vlef  %%v16,16(%%r1,%2),2        \n\t"
        "vlef  %%v17,20(%%r1,%2),2        \n\t"
        "vlef  %%v16,24(%%r1,%2),3        \n\t"
        "vlef  %%v17,28(%%r1,%2),3        \n\t"

Could be replaced with:
       "vl  %%v16,0(%%r1,%r2)                    \n\t"
       "vl  %%v1,16(%%r1,%r2)                    \n\t"
       "vpkg %%v17,%%v16%%v1               \n\t"
       "vperm %%v16,%%v16,%%v1,%%v0  \n\t"

However you would have to initialize v0 with {0,1,2,3,8,9,10,11,16,17,18,19,24,25,26,27}

**Comment by mtk-i** [comment]
Thanks for the tip, has been applied to my fork.

**Comment by mtk-i** [comment]
I created [PR#1991](https://github.com/xianyi/OpenBLAS/pull/1991) for this patch.

**Comment by martin-frbg** [comment]
Merged, thanks.

**Comment by aarnez** [comment]
Hm, OK... how to continue with this?  There are some things left to be cleaned up:
1. My patches need to be applied on top; otherwise OpenBLAS won't build on z14.
2. I haven't quite finished my testing of the complex min/max functions that aren't covered by BLAS-Tester. Findings may yet be uncovered.
3. The inline assemblies have bad constraints which just work by accident right now. This should be fixed.
4. I'd like the source code formatting to be made more consistent and readable.
So what's the best approach? Maybe I should start another patch set that addresses the remaining items?

**Comment by mtk-i** [comment]
oh alright, I will handle source code formatting and push it to my fork so you can preview, Also could you point me out to the bad constraints so I can take a look and potentially fix them.

**Comment by martin-frbg** [comment]
Sorry, my impression was that maamountki had included all changes that you suggested. I do not think there has been any harm done by merging the code even if it is still incomplete, just keep the patches coming.

**Comment by martin-frbg** [comment]
@aarnez if your patch from https://github.com/xianyi/OpenBLAS/issues/992#issuecomment-451257664 is still valid, I can also apply it directly if you like ?

**Comment by aarnez** [comment]
@maamountki:
> oh alright, I will handle source code formatting and push it to my fork so you can preview,

Great, thanks! It would be nice to make the indentation consistent, to remove trailing whitespace and unecessary empty lines, and to limit the line width to 80 chars. Maybe the `indent` tool can help, using options like `--no-space-after-function-call-names --dont-break-procedure-type --braces-on-if-line --braces-on-func-def-line --swallow-optional-blank-lines -T BLASLONG -T FLOAT`.

>  Also could you point me out to the bad constraints so I can take a look and potentially fix them.

In many instances the patch uses the `ZR` constraint when an address is needed in a register. This probably just works by accident right now. Use `a` instead. In order to declare the underlying object (and not just its address) as input, add an extra `m`-constraint input for it.

In addition:
* Please remove the `volatile` keyword, since it shouldn't be needed in these cases. If it makes a difference, something is likely broken.
* Similarly, remove the `"memory"` clobber. These inline assemblies should only write objects that are declared as outputs.

**Comment by aarnez** [comment]
@martin-frbg:
> if your patch from [#992 (comment)](https://github.com/xianyi/OpenBLAS/issues/992#issuecomment-451257664) is still valid, I can also apply it directly if you like ?

Better use this updated version: [Fixes-for-z14.txt](https://github.com/xianyi/OpenBLAS/files/2818745/Fixes-for-z14.txt). I just rebased it on develop, so it should apply cleanly. Thanks for your help!

**Comment by martin-frbg** [comment]
Applied and merged, thanks.

**Comment by aarnez** [comment]
@martin-frbg: Thanks for merging. It seems that you missed two patches, though:
* [PATCH 2/6] Fix incorrect sgemv results for IBM z14
* [PATCH 3/6] Fix typo in previous commit

Just applying them on top should be OK.

**Comment by martin-frbg** [comment]
Actually I appear to have misplaced the patched file in the PR as I used the web interface to upload it from a disconnected local repository - sgemv_t_4.c landed in the toplevel rather than kernel/zarch. Sigh.

**Comment by martin-frbg** [comment]
Should be sorted now in  874df65

**Comment by aarnez** [comment]
@martin-frbg: Looks OK now, thanks!

@maamountki: The remaining cleanup can now be done on top of current "develop". If you'd like some help with any of the cleanup work, please let me know.

**Comment by mtk-i** [comment]
@aarnez np, u seem busy, I can handle writing complex min/max functions tests that aren't covered by BLAS-Tester if u like.

**Comment by aarnez** [comment]
@maamountki, no need. I have some tests for those now; they should be sufficient for this purpose.

**Comment by mtk-i** [comment]
- Now the bad constraints are fixed and source code is formatted as suggested above.
- The constraints have names now as suggested by quickwritereader to be more readable and editable.
-While I was formatting the source code I addressed performance issues in gmev and ddot/sdot, They are faster now.
- Most kernels support up to 8 float multiplications and adds per cycle so Z13 and Z14 can take the advantage of the optimization despite Z13 can handle just 4, I'd suppose you will double vector registers count whenever you have new processor that can handle more than 8 float multiplications and adds per cycle.
- It's good to retest the whole thing now and once confirmed everything fine my work will be done on this issue.

**Comment by joncmu** [comment]
One other small performance comment.  On vl, vlm, vst, and vstm instructions when operating on double precision data the alignment hint value should be set to 3.  This may not show any performance improvement on z13 or z14 but may on a future machine.

**Comment by mtk-i** [comment]
Thanks, has been applied.

I just noticed I wrote above 8 instead of `8*2` for double precision and `8*4` for single precision, Likewise for 4.

**Comment by mtk-i** [comment]
@joncmu I didn't get it work I get `Error: junk at end of line` maybe it's not supported in Z13, Is it worth to separate kernels for applying this?

**Comment by mtk-i** [comment]
Conditional inclusions can solve the problem but I need confirmation that it works on Z14.

**Comment by aarnez** [comment]
@joncmu, @maamountki: It must be possible to build OpenBLAS with a binutils version that doesn't support alignment hints. Thus "vl" instructions cannot just be enriched with alignment hints unconditionally. As a circumvention, we can use `.insn` pseudo-opcodes instead, like this:
~~~
.insn vrx,0xe70000000006,%%v16,%[x],3
~~~
This should be supported by all binutils versions that support z13 insns. Maybe it helps to define a macro:
~~~
#define VL(operands) ".insn vrx,0xe70000000006," operands "\n\t"
~~~

**Comment by aarnez** [comment]
@maamountki, also note that the alignment hint must be specified correctly. For a general `float` array an alignment of 8 is not guaranteed, hence the alignment hint should be set to 0 instead of 3 in this case.

**Comment by mtk-i** [comment]
I agree, I also took a look to x86_64 implementations and the unaligned instructions seem to be used there.

**Comment by aarnez** [comment]
@maamountki, regarding the rework: Thanks for taking care of it, this has probably been quite an effort! For the most part, I like the changes. So far I've seen the following issues:

- The build fails when specifying `DEBUG=1`. This is because the constraints of some inline assemblies require too many registers when compiling with `-O0`. I've specifically investigated the function `sgemv_kernel_4x4()`, where this can be fixed by introducing variables `ap0` through `ap3` like this:
   ~~~
     register FLOAT *ap0 = ap[0];
     register FLOAT *ap1 = ap[1];
     register FLOAT *ap2 = ap[2];
     register FLOAT *ap3 = ap[3];
   ~~~
   and then using these variables instead of the array members as inputs for the inline assembly.
- The build works without `DEBUG=1`, but in my first tests I didn't see the expected performance. I might have done something wrong, though. Will repeat the test tomorrow.
- Many of the "m" constraints probably don't exactly do what you intended. E.g. the expression
   ~~~
   *(const FLOAT (*)[n]) ap0
   ~~~
  only dereferences the first array member, where you probably wanted the effect of this expression instead:
   ~~~
   *(const struct { FLOAT x[n]; } *) ap0
   ~~~
- In some cases the `"m"` constraint is used for an actual operand, e.g., for `alpha`. This should better be avoided and the appropriate choice of `"Q"`, `"R"`, `"S"`, or `"T"` should be used instead. Also, in the case of `alpha` it would probably be more efficient to pass the value instead, i.e., to specify `"f" (*alpha)` and receive the value in a floating-point register.
- The formatting is generally much better than before, but still a bit weird in some places. Did you use some tool for this? For instance, consider the indentation of inline assemblies:
   ~~~
     __asm__("vlrepf %%v0,0(%[x])\n\t"
          "vlrepf %%v1,4(%[x])\n\t"
          ...
   ~~~
   Here the second line is neither indented after the opening brace nor by the usual indentation width of 2. Anyway, this is just a minor nit, and I'm OK with leaving it as is.

**Comment by mtk-i** [comment]
I updated my fork with the changes you suggested.
- I used `indent` tool as you suggested, It's awesome tool but when it works on inline assembly it literally ruins it, It packs all the instructions in one line with only one space between them as well as it mess the constraints up so I found myself doing search & replace in trying to fix that but didn't go well anyway I fixed the formatting now in the inline assembly.
- I replaced the `"m"` constraint with "Q" where possible, I also investigated in replacing `"m"` with `"f"` for `alpha` and I think that inappropriate because that requires more instructions and for complex case even more registers since there are 2 variables for `alpha`.

**Comment by aarnez** [comment]
Great work, thanks!
Regarding the use of `"m"` for `alpha`: I'm OK with sticking to a memory reference, as long as we don't get "impossible constraint" errors -- and we don't. However, I'd really prefer "m" to be replaced by `"Q" (if you use instructions that don't support an index register) or `"R"` (if all used instructions support index/base and short displacement).
One problem still occurs with the current version: I see test failures due to wrong results from the single-precision complex functions cblas_cgemv, CGBMV, CGEMV, CGERC, CGERU, CHEMV, CHER, CHER2, CHPMV, CHPR, CHPR2, CTPMV, CTPSV, CTRMV, and CTRSV. If you need more info about these, let me know.

**Comment by mtk-i** [comment]
> 
> 
> However, I'd really prefer "m" to be replaced by `"Q" (if you use instructions that don't support an index register) or `"R"` (if all used instructions support index/base and short displacement).

even for those used to declare the underlying object?

**Comment by aarnez** [comment]
> even for those used to declare the underlying object?

No, it doesn't matter for operands that are not referenced in the inline assembly. It is actually a common pattern to use `"m"` in that case.

**Comment by mtk-i** [comment]
That's what I did previously, I pushed a fix for the test failures to my fork.

**Comment by aarnez** [comment]
Hm, your fix got rid of the CGEMV and CHEMV fails, but these calls still fail:
~~~
CHPMV ('U', 31,( 1.0, 0.0), AP, X, 1,( 0.0, 0.0), Y, 1)
CTPMV ('U','N','U', 31, AP, X, 1)
CTPSV ('U','N','U', 31, AP, X, 1)
CTRMV ('U','N','U', 31, A, 32, X, 1)
CTRSV ('U','N','U', 31, A, 32, X, 1)
CGERC ( 47, 31,( 1.0, 0.0), X, 1, Y, 1, A, 48)
CGERU ( 47, 31,( 1.0, 0.0), X, 1, Y, 1, A, 48)
CHER2 ('U', 31,( 1.0, 0.0), X, 1, Y, 1, A, 32)
CHPR2 ('U', 31,( 1.0, 0.0), X, 1, Y, 1, AP)
CHER  ('U', 31, 1.0, X, 1, A, 32)
CHPR  ('U', 31, 1.0, X, 1, AP)
~~~
And in the multi-threaded case all of the above fail as well, plus this one:
~~~
CGBMV ('T', 15, 31,  0,  0,( 0.7,-0.9), A,  2, X, 1,( 0.0, 0.0), Y, 1)
~~~
This problem must have been introduced by the cleanup patch, since there are no fails with the upstream version.

**Comment by mtk-i** [comment]
Should be fixed now.

**Comment by aarnez** [comment]
Great work! All test failures are gone now!

**Comment by aarnez** [comment]
OK, so what remains? A minor nit would be to apply the following change:
~~~
modified   kernel/zarch/cgemv_n_4.c
@@ -354,3 +354,3 @@ static void add_y_4()
     : [dest] "a"(dest), "m"(*(const struct { FLOAT x[n * 2]; } *) src),
-       [src] "a"(src),[alpha_r] "m"(alpha_r),[alpha_i] "m"(alpha_i)
+       [src] "a"(src),[alpha_r] "Q"(alpha_r),[alpha_i] "Q"(alpha_i)
     : "cc", "r1", "v0", "v1", "v16", "v17", "v18", "v19", "v20", "v21",
modified   kernel/zarch/zgemv_n_4.c
@@ -265,3 +265,3 @@ static void add_y_4()
     : [dest] "a"(dest), "m"(*(const struct { FLOAT x[n * 2]; } *) src),
-       [src] "a"(src),[alpha_r] "m"(alpha_r),[alpha_i] "m"(alpha_i)
+       [src] "a"(src),[alpha_r] "Q"(alpha_r),[alpha_i] "Q"(alpha_i)
     : "cc", "r1", "v0", "v1", "v16", "v17", "v18", "v19", "v20", "v21",
~~~
After that, I suggest to send a pull request. When that has made its way upstream, this issue can be closed.

**Comment by mtk-i** [comment]
Done, I created [PR#2012](https://github.com/xianyi/OpenBLAS/pull/2012)
@aarnez many thanks for your time, I enjoyed working with your assistance btw :-)

**Comment by martin-frbg** [comment]
PR #2012 has been merged into develop now.

**Comment by edelsohn** [comment]
Thanks for the great work. I believe that we're all in agreement that the feature request has been implemented and all pull requests landed.
